hardware_interface/doc/writing_new_hardware_interface.rst

```diff
@@ -0,0 +1,139 @@
+
+   4. Define the class of the hardware_interface, extending ``$InterfaceType$Interface``, e.g.,
+      .. code:: c++
+      class HardwareInterfaceName : public hardware_interface::$InterfaceType$Interface
+
+   5. Add a constructor without parameters and the following public methods implementing ``LifecycleNodeInterface``: ``on_configure``, ``on_cleanup``, ``on_shutdown``, ``on_activate``, ``on_deactivate``, ``on_error``; and overriding ``$InterfaceType$Interface`` definition: ``on_init``, ``export_state_interfaces``, ``export_command_interfaces``, ``prepare_command_mode_switch`` (optional), ``perform_command_mode_switch`` (optional), ``read``, ``write``.
+   For further explanation of hardware-lifecycle check the `pending pull request <https://github.com/ros-controls/ros2_control/pull/559/files#diff-2bd171d85b028c1b15b03b27d4e6dcbb87e52f705042bf111840e7a28ab268fc>`_ and for exact definitions of methods check the ``"hardware_interface/$interface_type$_interface.hpp"`` header or `doxygen documentation <http://control.ros.org/api/namespacehardware__interface.html>`_ for *Actuator*, *Sensor* or *System*.
+
+4. **Adding definitions into source file (.cpp)**
+
+   1. Include the header file of your hardware interface and add a namespace definition to simplify further development.
+
+   2. Implement ``on_init`` method. There, you should all initialize member variables and process hardware's parameters from ``info`` argument.
+      In the first line usually the parents ``on_init`` is called to process standard values, like name. This is done using: ``hardware_interface::(Actuator|Sensor|System)Interface::on_init(info)``.
+      If all required parameters are set and valid and everything works fine return ``CallbackReturn::SUCCESS`` or ``return CallbackReturn::ERROR`` otherwise.
+
+   4. Write the ``on_configure`` method where you usually setup the communication to the hardware and set everything up so that the hardware can be activated.
+
+   5. Implement ``on_cleanup`` method, which does the opposite of ``on_configure``.
+
+   6. Implement ``export_state_interfaces`` and ``export_command_interfaces`` where interfaces that hardware if offering are defined.
+      For the ``Sensor``-type hardware interface there is no ``export_command_interfaces`` method.
+      As a reminder, the full interface names have structure ``<joint_name>/<interface_type>``.
+
+   7. (optional) For *Actuator* and *System* types of hardware interface implement ``prepare_command_mode_switch`` and ``perform_command_mode_switch`` if your hardware accepts multiple control modes.
+
+   8. Implement the ``on_activate`` method where hardware "power" is enabled.
+
+   9. Implement the ``on_deactivate`` method, which does the opposite of ``on_activate``.
+
+   10. Implement ``on_shutdown`` method where hardware is shutdown gracefully.
+
+   11. Implement ``on_error`` method where different errors from all states are handled.
```

Comment:
we could add something about recovering from errors being possible here?

Reply:
Probably we should link the explanation from [here](http://control.ros.org/ros2_control/hardware_interface/doc/hardware_components_userdoc.html#handling-of-errors-that-happen-during-read-and-write-calls). What do you think.